### PR TITLE
feat: [CDS-71893]: Handle resolution of objects via json select behin…

### PR DESCRIPTION
**Pull request by abhinavmittal-harness**
JSON FUNCTOR used in http step was the one we used in CG.
And the select of that only used to resolve either String attributes or a list of size 1.
This is limiting in terms of usages.
So we have decided to expand it to all type of objects behind a feature flag.

Images of tests run locally added
Screenshot 2023-06-12 at 8 49 18 PM

**With feature flag on**

Screenshot 2023-06-12 at 8 49 02 PM
image
Screenshot 2023-06-19 at 5 39 35 PM

**With ff off**

Screenshot 2023-06-12 at 8 49 27 PM
image
Screenshot 2023-06-19 at 5 39 16 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49122)

### CODE REVIEW

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 1]
```diff
@@ -806,7 +806,9 @@ public enum FeatureName {
       PIPELINE),
 
   SRM_ENABLE_SLI_BUCKET("This is used to enable sli bucket reads", HarnessTeam.CV),
-  PL_FAVORITES("To enable favorites marking support on entities", HarnessTeam.PL);
+  PL_FAVORITES("To enable favorites marking support on entities", HarnessTeam.PL),
+  RESOLVE_OBJECTS_VIA_JSON_SELECT(
```
Comment (puthrayaharness) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:810]: Please prefix with `CDS_`

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

**Review by archit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49122)* status: 4 file(s) to review. 2 change requests. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/AmbianceExpressionEvaluator.java` line 136 at r1](https://reviewable.io/reviews/harness/harness-core/49122#-NXo9wQT0zxSwcGsd7Ss:-NXo9wQT0zxSwcGsd7St:bracna0) ([raw file](https://github.com/harness/harness-core/blob/37d56858eef5fef9c9d1dcc2e6e8fbe2c82491b5/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/AmbianceExpressionEvaluator.java#L136)):*
> ```Java
>         boolean resolveObjectsViaJSONSelect;
>         if (pmsFeatureFlagService.isEnabled(
>                 AmbianceUtils.getAccountId(ambiance), FeatureName.RESOLVE_OBJECTS_VIA_JSON_SELECT)) {
> ```

As client is sending contextMap, they can add a key which is true or false, which can be used here and not FF directly

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49122)* status: 6 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/expression/JsonFunctor.java` line 44 at r3](https://reviewable.io/reviews/harness/harness-core/49122#-NYH1jBZ3yKNtIGP09Rm:-NYH1jBZ3yKNtIGP09Rn:b-2kcs6b) ([raw file](https://github.com/harness/harness-core/blob/55a1d8744691d5215cd83f1fe32659d637dd6bc8/980-commons/src/main/java/io/harness/expression/JsonFunctor.java#L44)):*
> ```Java
>       }
>     }
>     return contextMap != null && contextMap.containsKey("resolveObjectsViaJSONSelect") ? object : null;
> ```

Can you add constant and use it here and in HttpStep?



_Code quote:_
```Java
"resolveObjectsViaJSONSelect"
```
___
*[`980-commons/src/test/java/io/harness/expression/JsonFunctorTest.java` line 25 at r3](https://reviewable.io/reviews/harness/harness-core/49122#-NYH1oqs2P5OuRoyLyiS:-NYH1oqs2P5OuRoyLyiT:bgu7fv2) ([raw file](https://github.com/harness/harness-core/blob/55a1d8744691d5215cd83f1fe32659d637dd6bc8/980-commons/src/test/java/io/harness/expression/JsonFunctorTest.java#L25)):*
> ```Java
> import org.junit.experimental.categories.Category;
> 
> public class JsonFunctorTest extends CategoryTest {
> ```

Add team annotation

**Review by abhinavmittal-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49122)* status: 11 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/expression/JsonFunctor.java` line 44 at r3](https://reviewable.io/reviews/harness/harness-core/49122#-NYH1jBZ3yKNtIGP09Rm:-NYHwfmL4Ni5oEKtTsUE:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/55a1d8744691d5215cd83f1fe32659d637dd6bc8/980-commons/src/main/java/io/harness/expression/JsonFunctor.java#L44)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Can you add constant and use it here and in HttpStep?
</blockquote></details>

Done.
___
*[`980-commons/src/test/java/io/harness/expression/JsonFunctorTest.java` line 25 at r3](https://reviewable.io/reviews/harness/harness-core/49122#-NYH1oqs2P5OuRoyLyiS:-NYHwgSk22nNkMYZP9cc:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/55a1d8744691d5215cd83f1fe32659d637dd6bc8/980-commons/src/test/java/io/harness/expression/JsonFunctorTest.java#L25)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Add team annotation
</blockquote></details>

Done.

**Comment by abhinavmittal-harness** [comment]
trigger uts

**Review by archit-harness — DISMISSED** [review]
> Reviewed 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49122)* status: 11 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by puthrayaharness — APPROVED** [review]

**Comment by abhinavmittal-harness** [comment]
trigger review

**Comment by yogesh-chauhan** [comment]
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 284 at r4](https://reviewable.io/reviews/harness/harness-core/49122#-NYIC1964GsuNukXPRna:-NYIC1964GsuNukXPRnb:b-ibjd3k) ([raw file](https://github.com/harness/harness-core/blob/47437517f6624f52671405885e89ae6e769f6673/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L284)):*
> ```Java
>     } else {
>       resolveObjectsViaJSONSelect = false;
>     }
> ```

This can be made less verbose and also made to use ```final```
```
final boolean resolveObjectsViaJSONSelect = pmsFeatureFlagHelper.isEnabled(
            AmbianceUtils.getAccountId(ambiance), FeatureName.CDS_RESOLVE_OBJECTS_VIA_JSON_SELECT)

```

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 6 files at r2, 2 of 2 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49122)* status: 1. unresolved discussions.

**Comment by abhinavmittal-harness** [comment]
trigger ut3

**Review by abhinavmittal-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49122)* status: 1 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java` line 284 at r4](https://reviewable.io/reviews/harness/harness-core/49122#-NYIC1964GsuNukXPRna:-NYIVvw9D9aRV7k3w_cE:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/47437517f6624f52671405885e89ae6e769f6673/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/http/HttpStep.java#L284)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

This can be made less verbose and also made to use ```final```
```
final boolean resolveObjectsViaJSONSelect = pmsFeatureFlagHelper.isEnabled(
            AmbianceUtils.getAccountId(ambiance), FeatureName.CDS_RESOLVE_OBJECTS_VIA_JSON_SELECT)

```

</blockquote></details>

Done.

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49122)* status: 1 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Comment by abhinavmittal-harness** [comment]
trigger uts

**Comment by abhinavmittal-harness** [comment]
trigger review

**Comment by abhinavmittal-harness** [comment]
trigger ut9

**Pull request merged**